### PR TITLE
[EXTERNAL PR] TTNN Roll Op Support

### PR DESCRIPTION
**Pull request by Bhuvanesh194Sankar**
### #12157 
Link to Github Issue : https://github.com/tenstorrent/tt-metal/issues/12157

### Problem description
Need support for ttnn roll() op.

### What's changed
Implemented roll op (ttnn.roll) as a composite op using the ttnn.slice and ttnn.concat, and added pytest .

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI [passes](https://github.com/tenstorrent/tt-metal/actions/runs/15045391672/job/42312144550)
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Comment by Bhuvanesh194Sankar** [comment]
Hello team, can someone kindly review this PR ?

File: tests/ttnn/unit_tests/test_roll.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import pytest
+import torch
+import ttnn
+from tests.ttnn.utils_for_testing import assert_with_pcc
+
+
+@pytest.mark.parametrize(
+    "input_tensor, shifts, dim",
+    [
+        ([[1, 2], [3, 4], [5, 6], [7, 8]], 1, 0),
```
Comment (bbradelTT) [inline tests/ttnn/unit_tests/test_roll.py:14]: Is dim being None supported?
- If it is supported, please add a test.
- If it isn't supported, should it be supported?
  - If it should be, then the code needs to be added (in this or a subsequent PR).
  - If it shouldn't be supported, please note that in the doc string.

File: ttnn/cpp/ttnn/operations/data_movement/roll/roll.cpp [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+#include "ttnn/operations/core/core.hpp"
+#include "cpp/ttnn/operations/data_movement/slice/slice.hpp"
+#include "cpp/ttnn/operations/data_movement/concat/concat.hpp"
+#include "ttnn/tensor/tensor.hpp"
+
+namespace ttnn::operations::data_movement {
+
+ttnn::Tensor RollOperation::invoke(
+    const ttnn::Tensor& input_tensor, const std::vector<int>& shifts, const std::vector<int>& dims) {
+    TT_FATAL(!shifts.empty() && shifts.size() == dims.size(), "Roll expects shifts and dims to have the same length");
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/data_movement/roll/roll.cpp:15]: Please print out values in the error statement.

File: ttnn/cpp/ttnn/operations/data_movement/roll/roll.cpp [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        int shift_size = input_tensor.get_logical_shape()[dim];
+        adjusted_shifts[i] = ((shift % shift_size) + shift_size) % shift_size;
+    }
+
+    ttnn::Tensor result = input_tensor;
+    auto size = result.get_logical_shape();
+    int num_dims = size.rank();
+
+    for (int dim : dims) {
+        TT_FATAL(dim >= -num_dims && dim < num_dims, "Invalid dimension index.");
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/data_movement/roll/roll.cpp:32]: Please print out values in the error statement.

**Review by bbradelTT — APPROVED** [review]

**Comment by Bhuvanesh194Sankar** [comment]
@bbradelTT Thank you for your comments. I will work on them and update the changes at the earliest

File: tests/ttnn/unit_tests/test_roll.py (hunk 1 above)
Comment (Bhuvanesh194Sankar) [reply tests/ttnn/unit_tests/test_roll.py:14]: Will add support for dim being None in this pr itself.

File: ttnn/cpp/ttnn/operations/data_movement/roll/roll.cpp [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    TT_FATAL(!shifts.empty() && shifts.size() == dims.size(), "Roll expects shifts and dims to have the same length");
+
+    std::vector<int> adjusted_shifts = shifts;
+
+    for (size_t i = 0; i < adjusted_shifts.size(); ++i) {
+        int shift = adjusted_shifts[i];
+        int dim = dims[i];
+
+        int shift_size = input_tensor.get_logical_shape()[dim];
+        adjusted_shifts[i] = ((shift % shift_size) + shift_size) % shift_size;
```
[the remark this answers is not in the document]
Comment (Bhuvanesh194Sankar) [reply ttnn/cpp/ttnn/operations/data_movement/roll/roll.cpp:24]: @bbradelTT 
adjusted_shifts[i] = (shift + shift_size) % shift_size; -> This code snippet that you have suggested will work on positive shift, 
adjusted_shifts[i] = ((shift % shift_size) + shift_size) % shift_size; -> This will handle both postive and negative shifts aswell.
So to handle the negative shifts our code snippet is necessary.

File: tests/ttnn/unit_tests/test_roll.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        ),
+        ([[[1, 2], [3, 4]], [[5, 6], [7, 8]], [[9, 10], [11, 12]]], [1, 0], [0, 1]),
+        ([[[[1, 2], [3, 4]], [[5, 6], [7, 8]]], [[[9, 10], [11, 12]], [[13, 14], [15, 16]]]], [1, 0, -2], [0, 1, 2]),
+        ([[[[1, 2], [3, 4]], [[5, 6], [7, 8]], [[9, 10], [11, 12]], [[13, 14], [15, 16]]]], [0, -1, -2], [0, 1, 2]),
+    ],
+)
+def test_roll(device, input_tensor, shifts, dim):
+    tensor = torch.tensor(input_tensor, dtype=torch.float32)
+    ttnn_tensor = ttnn.from_torch(tensor)
+    ttnn_tensor = ttnn.to_layout(ttnn_tensor, ttnn.ROW_MAJOR_LAYOUT)
```
Comment (ntarafdar) [inline tests/ttnn/unit_tests/test_roll.py:45]: can we test for TILE and other dtypes (bfloat16 in particular) as well. 

Also for TILE can we try shapes that are not divisible by 32 for the last two dims to see how this would handle padding. 

It should be okay as it uses slice and concat underneath the hood which has this support. 

Can we also try L1 memory and diff sharding configs as well. 

And lastly can we test the program cache, (use_program_cache and have multiple runs)

**Review by ntarafdar — CHANGES_REQUESTED** [review]
just a small comment on testing.

File: tests/ttnn/unit_tests/test_roll.py (hunk 1 above)
Comment (Bhuvanesh194Sankar) [reply tests/ttnn/unit_tests/test_roll.py]: @bbradelTT I have added the None dim support can you kindly review it ?

**Comment by Bhuvanesh194Sankar** [comment]
@ntarafdar I have done the suggested changes, can you kindly review it again ?

File: tests/ttnn/unit_tests/test_roll.py (hunk 1 above)
Comment (bbradelTT) [reply tests/ttnn/unit_tests/test_roll.py]: Where are the tests for the None dim?

File: tests/ttnn/unit_tests/test_roll.py [hunk 6]
```diff
@@ -33,21 +33,193 @@
...
+            ],
+            -2,
+        ),
+        ([[[1, 2], [3, 4]], [[5, 6], [7, 8]], [[9, 10], [11, 12]]], 1),
+        ([[[[1, 2], [3, 4]], [[5, 6], [7, 8]]], [[[9, 10], [11, 12]], [[13, 14], [15, 16]]]], 1),
+        ([[[[1, 2], [3, 4]], [[5, 6], [7, 8]], [[9, 10], [11, 12]], [[13, 14], [15, 16]]]], -1),
     ],
 )
-def test_roll(device, input_tensor, shifts, dim):
+def test_roll_without_dim(device, input_tensor, shifts):
```
Comment (Bhuvanesh194Sankar) [inline tests/ttnn/unit_tests/test_roll.py:113]: @bbradelTT here is the test for roll without dim

Comment (bbradelTT) [reply tests/ttnn/unit_tests/test_roll.py:113]: Thanks.

**Comment by Bhuvanesh194Sankar** [comment]
@bbradelTT removed the print statements from the test fucntions can you kindly review it.

File: ttnn/cpp/ttnn/operations/data_movement/roll/roll.cpp [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+        ttnn::Tensor right_part = ttnn::slice(result, start_right, end_right, ttnn::SmallVector<int>(num_dims, 1));
+
+        std::vector<ttnn::Tensor> tensors_to_concat = {left_part, right_part};
+        result = ttnn::concat(tensors_to_concat, dim);
+    }
+
+    return result;
+}
+
+ttnn::Tensor RollOperation::invoke(const ttnn::Tensor& input_tensor, const int& shift) {
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/data_movement/roll/roll.cpp:81]: This is too much code duplication. There should be one method that does the operations (like in the first commit).
- Either this is an internal method and all three invoke methods call that one.
- Or it is the initial invoke method with the other two invoke methods (this one and the one below) calling the initial invoke method.

File: ttnn/cpp/ttnn/operations/data_movement/roll/roll_pybind.cpp [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        ttnn::roll,
+        R"doc(
+        roll(input_tensor: ttnn.Tensor, shifts: Union[int, List[int]], dim: Union[int, List[int]]) -> ttnn.Tensor
+
+        Performs circular shifting of elements along the specified dimension(s).
+
+        Args:
+            input_tensor: A tensor whose elements will be rolled.
+            shifts: The number of places by which elements are shifted. Can be an integer or a list of integers (one per dimension).
+            dim: The dimension(s) along which to roll. Must have the same length as `shifts` if a list is provided.
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/data_movement/roll/roll_pybind.cpp:63]: This is not clear if the list is for shifts or dim. 
The end should probably be changed to
"If `shifts` is a list, then `dim` must be a list of the same length as `shifts`."

**Comment by Bhuvanesh194Sankar** [comment]
Kindly share if there are any other review comments. If not, could anyone please trigger the CIs for the PR? Thank you. Also, @ntarafdar  , your approval might be blocking the PR. Could you approve the PR, if no further comments from your side?

**Comment by bbradelTT** [comment]
> @bbradelTT Kindly share if there are any other review comments. If not, could you please trigger the CIs for the PR? Thank you. Also, @ntarafdar , your approval might be blocking the PR. Could you approve the PR, if no further comments from your side?

Thanks @Bhuvanesh194Sankar the changes look good.

I think you still need to address the comment from @ntarafdar about the tests https://github.com/tenstorrent/tt-metal/pull/19676/files#r2039494468 and then he can trigger the CIs.

**Comment by Bhuvanesh194Sankar** [comment]
> > @bbradelTT Kindly share if there are any other review comments. If not, could you please trigger the CIs for the PR? Thank you. Also, @ntarafdar , your approval might be blocking the PR. Could you approve the PR, if no further comments from your side?
> 
> Thanks @Bhuvanesh194Sankar the changes look good.
> 
> I think you still need to address the comment from @ntarafdar about the tests https://github.com/tenstorrent/tt-metal/pull/19676/files#r2039494468 and then he can trigger the CIs.

Thank you @bbradelTT.  Regarding the comment from @ntarafdar, I have already addressed it and added the suggested tests in the test_roll.py test file.

**Comment by Bhuvanesh194Sankar** [comment]
@bbradelTT @ntarafdar The tests for the different sharding configs were not included because this roll op makes use of slice and concat, the slice supports sharding but concat has some constraints on the sharding, which throws the fatals, the fatals in the code is listed below as the images for your reference, these fatals are in the file 
tt-metal/ttnn/cpp/ttnn/operations/data_movement/concat/device/concat_device_operation.cpp

image

image

So other than that I have addressed all the other comments for test suggested by @ntarafdar which was in the test_roll.py test python file.

**Comment by Bhuvanesh194Sankar** [comment]
@bbradelTT @ntarafdar  Could some one review this ?

**Review by ntarafdar — APPROVED** [review]

**Comment by Bhuvanesh194Sankar** [comment]
@ntarafdar  Thank you for the approval, the CI runs triggered yesterday seemed fine. I have rebased the branch to the latest main. Could you please trigger the CI runs again? If they pass can we proceed with merge?

**Comment by Bhuvanesh194Sankar** [comment]
@amorrisonTT Could you kindly review this PR  ?

**Comment by Bhuvanesh194Sankar** [comment]
@amorrisonTT Could you kindly review this PR ?

File: ttnn/cpp/ttnn/operations/data_movement/roll/roll.cpp [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+#include "ttnn/operations/core/core.hpp"
+#include "cpp/ttnn/operations/data_movement/slice/slice.hpp"
+#include "cpp/ttnn/operations/data_movement/concat/concat.hpp"
+#include "ttnn/tensor/tensor.hpp"
+#include "ttnn/operations/data_movement/reshape_view/reshape.hpp"
+
+namespace ttnn::operations::data_movement {
+
+ttnn::Tensor RollOperation::invoke(
+    const ttnn::Tensor& input_tensor, const ttnn::SmallVector<int>& shift, const ttnn::SmallVector<int>& dim) {
```
Comment (amorrisonTT) [inline ttnn/cpp/ttnn/operations/data_movement/roll/roll.cpp:15]: I think you need to rename the input argument `dim` since it gets shadowed in the function body. I would suggest:  `input_dims`

**Comment by Bhuvanesh194Sankar** [comment]
@amorrisonTT I have done the suggested changes, could you kindly review it again ?

**Comment by Bhuvanesh194Sankar** [comment]
@amorrisonTT We can proceed with the CI runs .

**Comment by tt-rkim** [comment]
running https://github.com/tenstorrent/tt-metal/actions/runs/14910320925/job/41884018121?pr=19676

**Comment by Bhuvanesh194Sankar** [comment]
@tt-rkim it seems five checks were failing. What should we do next ?

**Comment by Bhuvanesh194Sankar** [comment]
@tt-rkim I have rebased the branch to the latest main, can we go for another CI run?

**Comment by Bhuvanesh194Sankar** [comment]
@tt-rkim I have rebased the branch to the latest main. Could you please trigger the CI runs again ?

**Review by tt-rkim — APPROVED** [review]

**Comment by tt-rkim** [comment]
APC here https://github.com/tenstorrent/tt-metal/actions/runs/14995088226/job/42132248644?pr=19676

File: ttnn/cpp/ttnn/operations/data_movement/roll/roll.cpp [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+#include "cpp/ttnn/operations/data_movement/concat/concat.hpp"
+#include "ttnn/tensor/tensor.hpp"
+#include "ttnn/operations/data_movement/reshape_view/reshape.hpp"
+
+namespace ttnn::operations::data_movement {
+
+ttnn::Tensor RollOperation::invoke(
+    const ttnn::Tensor& input_tensor, const ttnn::SmallVector<int>& shift, const ttnn::SmallVector<int>& input_dims) {
+    std::vector<int> shifts(shift.begin(), shift.end());
+    std::vector<int> dims(input_dims.begin(), input_dims.end());
```
Comment (sminakov-tt) [inline ttnn/cpp/ttnn/operations/data_movement/roll/roll.cpp:17]: It seems like this conversion from SmallVector to std::vector is redundant, can we use SmallVector throughout this function?

**Review by sminakov-tt — APPROVED** [review]

**Comment by tt-rkim** [comment]
Note that other people can run it as well, and someone already triggered it. Please refer to the statuses below on this page, before rebasing again, as that will erase it.
